### DISCUSSION TITLE
Duplicate detection fails when tags are changed

### DISCUSSION THREAD

**Discussion by mzeising** [firefly-iii/firefly-iii#6747] [Firefly III]
### Support guidelines

- [X] I've read the [support guidelines](https://github.com/firefly-iii/firefly-iii/blob/main/.github/support.md)

### I've found a bug and checked that ...

- [X] ... [the documentation](https://docs.firefly-iii.org/) does not mention anything about my problem
- [X] ... there are no open or closed issues that are related to my problem
- [X] ... it's [definitely me, not you](https://github.com/firefly-iii/firefly-iii/blob/main/.github/its_you_not_me.md)

### Description

The hash for a transaction includes its tags. This results in transactions not being identified as duplicates when their tags are changed.

### Expected behaviour

Exclude tags from the hash for transactions

### Debug information

(there is no debug information)

### Additional info

As the *Firefly III Data Importer* suggests, I use tags to denote identify import runs. Each import run uses a unique tag like *"Import no. 3 on 01/12/22"*. Obviously, duplicates are never found this way.

**Comment by JC5** [comment]
This is by design. Everything you submit to Firefly III is part of the hash. Notable exceptions are the import hash itself (to prevent recursion) and if present, the 'original source' field that may change between data importer versions.

If you need a more specific way to detect duplicates, you could use "identifier-based detection" and select a unique field amongst your transactions, such as an external ID.

**Comment by mzeising** [comment]
- How can I change the detection to identifier-based?
- The main purpose of using tags is that I want to be able to undo an import. As far as I understand, this is not possible using the field *original source*, is it?
- Isn't this a big issue for FIDI? It adds a date-based tag as well, doesn't it?

**Reply by JC5** [reply]
It's a dropdown in the configuration page. You can use the resulting JSON file to change your own if necessary (I believe the value is `cell`, + an index field + the name of the field where the value is stored).

Like you say, using tags is an excellent idea to undo imports. The data importer cheats by adding the tag *after* the import. It loops all transactions it added and then adds the tag. This keeps the tag out of the duplicate check, because the hash is not updated when you change transactions.

 Keep in mind that deleted transactions will still trigger the duplicate detection.

**Reply by mzeising** [reply]
I'm sorry, I can't find that dropdown, are you refering to the data importer's UI? I'm not using it, I'm loading transactions per Firefly's API directly.

**Reply by JC5** [reply]
Then you'll have to tinker something together yourself. If you use the identifier-based detection method in the data importer, the data importer runs a query before each transaction to see if it exists (`external_id:"abc..."`). 

You can replicate this behavior yourself.
